wallet: Avoid translating RPC errors

maflcko opened the pull request:
Common errors and warnings should be translated when displayed in the
GUI, but not translated when displayed elsewhere. The wallet method
`CreateWalletFromFile` does not know its caller, so this commit changes it
to return a `bilingual_str` to the caller.

Fixes #17072

Commit:
wallet: Avoid translating RPC errors when loading wallets

Common errors and warnings should be translated when displayed in the
GUI, but not translated when displayed elsewhere. The wallet method
CreateWalletFromFile does not know its caller, so this commit changes it
to return a bilingual_str to the caller.

MarcoFalke commented:
Can be tested with:

```
XDG_SESSION_TYPE=""  QT_QPA_PLATFORM=minimal LC_ALL=de_DE.UTF-8 BITCOIND=bitcoin-qt ./test/functional/wallet_basic.py 
```

The gui popups should be translated, but the json rpc errors should not.

Commit:
wallet: Avoid translating RPC errors when creating txs

Also, mark feebumper bilingual_str as Untranslated

They are technical and have previously not been translated either.
It is questionable whether they can even appear in the GUI.

hebasto commented:
Concept ACK.

Commit:
wallet: Report full error message in wallettool

MarcoFalke commented:
Rebased

hebasto commented:
@MarcoFalke 


> Can be tested with:
> 
> ```
> XDG_SESSION_TYPE=""  QT_QPA_PLATFORM=minimal LC_ALL=de_DE.UTF-8 BITCOIND=bitcoin-qt ./test/functional/wallet_basic.py 
> ```
> 
> The gui popups should be translated, but the json rpc errors should not.

On Linux Mint 19.3 having a timeout:
```
$ QT_LOGGING_RULES="qt5ct.debug=false" XDG_SESSION_TYPE="" QT_QPA_PLATFORM=minimal LC_ALL=ru_RU.UTF-8 BITCOIND=/home/hebasto/GitHub/bitcoin/src/qt/bitcoin-qt ./test/functional/wallet_basic.py
2020-04-21T10:25:52.532000Z TestFramework (INFO): Initializing test directory /tmp/bitcoin_func_test_ya2ujqgn
2020-04-21T10:25:53.484000Z TestFramework (INFO): Mining blocks...
2020-04-21T10:26:54.210000Z TestFramework (ERROR): Assertion failed
Traceback (most recent call last):
  File "/home/hebasto/GitHub/bitcoin/test/functional/test_framework/test_framework.py", line 112, in main
    self.run_test()
  File "./test/functional/wallet_basic.py", line 65, in run_test
    self.sync_all(self.nodes[0:3])
  File "/home/hebasto/GitHub/bitcoin/test/functional/test_framework/test_framework.py", line 490, in sync_all
    self.sync_blocks(nodes, **kwargs)
  File "/home/hebasto/GitHub/bitcoin/test/functional/test_framework/test_framework.py", line 484, in sync_blocks
    sync_blocks(nodes or self.nodes, **kwargs)
  File "/home/hebasto/GitHub/bitcoin/test/functional/test_framework/util.py", line 413, in sync_blocks
    raise AssertionError("Block sync timed out:{}".format("".join("\n  {!r}".format(b) for b in best_hash)))
AssertionError: Block sync timed out:
  '25fa4cf68bbf80d0b9cf47795f8eff272b7230d16c94f0b1a82602aadf6398a2'
  '0f9188f13cb7b2c71f2a335e3a4fc328bf5beb436012afca590b1a11466e2206'
  '0f9188f13cb7b2c71f2a335e3a4fc328bf5beb436012afca590b1a11466e2206'
2020-04-21T10:26:54.266000Z TestFramework (INFO): Stopping nodes
2020-04-21T10:26:54.726000Z TestFramework (WARNING): Not cleaning up dir /tmp/bitcoin_func_test_ya2ujqgn
2020-04-21T10:26:54.726000Z TestFramework (ERROR): Test failed. Test logging available at /tmp/bitcoin_func_test_ya2ujqgn/test_framework.log
2020-04-21T10:26:54.727000Z TestFramework (ERROR): Hint: Call /home/hebasto/GitHub/bitcoin/test/functional/combine_logs.py '/tmp/bitcoin_func_test_ya2ujqgn' to consolidate all logs
```

hebasto commented:
Should `CheckUniqueFileid()` errors be translated for GUI (possibly in other PR)?

hebasto reviewed (approved):
ACK fa5ba7e472b90707f7ae13ba2ab9e2b106a9c743, tested on Linux Mint 19.3 by manually running commands in the GUI and in the console window.

Noted that it is not possible to confirm that _all_ touched in this PR error messages are translated in the GUI as translation files are not updated for now, e.g., https://github.com/bitcoin/bitcoin/blob/fadc12f29e17451a090fc7fa03bdf4603ab874ba/src/wallet/wallet.cpp#L157

MarcoFalke commented:
> Noted that it is not possible to confirm that all touched in this PR error messages are translated in the GUI as translation files are not updated for now, e.g., 

This is always true, as translations may be missing. You reminded me to remove trailing whitespace from translations, since whitespace is hard to translate and easy to forget. Added a new commit, no changes to the commit you already reviewed.

MarcoFalke commented:
> On Linux Mint 19.3 having a timeout:

Is there an error popup? You might have to close it. You can also try without the `QT_QPA_PLATFORM` and manually close all error popups.

hebasto commented:
> > On Linux Mint 19.3 having a timeout:
> 
> Is there an error popup? You might have to close it. You can also try without the `QT_QPA_PLATFORM` and manually close all error popups.

No error popups. Without the `QT_QPA_PLATFORM` I can see three windows but no error popups.

hebasto commented on src/wallet/feebumper.cpp:
```diff
@@ -218,9 +220,9 @@ Result CreateRateBumpTransaction(CWallet& wallet, const uint256& txid, const CCo
+    bilingual_str fail_reason;
     if (!wallet.CreateTransaction(*locked_chain, recipients, tx_new, fee_ret, change_pos_in_out, fail_reason, new_coin_control, false)) {
-        errors.push_back("Unable to create transaction: " + fail_reason);
+        errors.push_back(_("Unable to create transaction.") + fail_reason);
```
fa5a95c678ba2a4ebdbaad084a9fe18ba5b64629
Hereinafter, it should looks like https://github.com/bitcoin/bitcoin/blob/fa5a95c678ba2a4ebdbaad084a9fe18ba5b64629/src/wallet/wallet.cpp#L199

MarcoFalke replied:
Thanks, good catch.

Commit:
wallet: Remove trailing whitespace from potential translation strings

If the potential translation strings are translated in the future,
trailing whitespace is going to make translation effort harder.

hebasto commented:
On Ubuntu 18.04.4 having another kind of exception:
```
$ XDG_SESSION_TYPE="" QT_QPA_PLATFORM=minimal LC_ALL=de_DE.UTF-8 BITCOIND=/home/hebasto/bitcoin/src/qt/bitcoin-qt ./test/functional/wallet_basic.py
2020-04-21T12:15:56.635000Z TestFramework (INFO): Initializing test directory /tmp/bitcoin_func_test_r5a4hpii
2020-04-21T12:16:16.255000Z TestFramework (INFO): Mining blocks...
2020-04-21T12:16:17.987000Z TestFramework (INFO): test gettxout
2020-04-21T12:16:21.593000Z TestFramework (INFO): test gettxout (second part)
--- Logging error ---
Traceback (most recent call last):
  File "/home/hebasto/bitcoin/test/functional/test_framework/test_framework.py", line 112, in main
    self.run_test()
  File "./test/functional/wallet_basic.py", line 395, in run_test
    test_address(self.nodes[0], addr, labels=[label])
  File "/home/hebasto/bitcoin/test/functional/test_framework/wallet_util.py", line 99, in test_address
    raise AssertionError("key {} value {} did not match expected value {}".format(key, addr_info[key], value))
AssertionError: key labels value [{'name': '\u0440\u044b\u0431\u0430', 'purpose': 'receive'}] did not match expected value ['\u0440\u044b\u0431\u0430']

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib/python3.6/logging/__init__.py", line 996, in emit
    stream.write(msg)
UnicodeEncodeError: 'ascii' codec can't encode characters in position 600-603: ordinal not in range(128)
Call stack:
  File "./test/functional/wallet_basic.py", line 536, in <module>
    WalletTest().main()
  File "/home/hebasto/bitcoin/test/functional/test_framework/test_framework.py", line 120, in main
    self.log.exception("Assertion failed")
Message: 'Assertion failed'
Arguments: ()
2020-04-21T12:17:41.175000Z TestFramework (INFO): Stopping nodes
2020-04-21T12:17:48.063000Z TestFramework (WARNING): Not cleaning up dir /tmp/bitcoin_func_test_r5a4hpii
2020-04-21T12:17:48.063000Z TestFramework (ERROR): Test failed. Test logging available at /tmp/bitcoin_func_test_r5a4hpii/test_framework.log
2020-04-21T12:17:48.065000Z TestFramework (ERROR): Hint: Call /home/hebasto/bitcoin/test/functional/combine_logs.py '/tmp/bitcoin_func_test_r5a4hpii' to consolidate all logs
```

MarcoFalke commented:
Did you set `PYTHONUTF8=1`?

hebasto commented:
> Did you set `PYTHONUTF8=1`?

```
$ PYTHONUTF8=1 XDG_SESSION_TYPE="" QT_QPA_PLATFORM=minimal LC_ALL=de_DE.UTF-8 BITCOIND=/home/hebasto/bitcoin/src/qt/bitcoin-qt ./test/functional/wallet_basic.py
```
didn't help ((

hebasto commented:
> Can be tested with:
> 
> ```
> XDG_SESSION_TYPE=""  QT_QPA_PLATFORM=minimal LC_ALL=de_DE.UTF-8 BITCOIND=bitcoin-qt ./test/functional/wallet_basic.py 
> ```
> 
> The gui popups should be translated, but the json rpc errors should not.

On Ubuntu 18.04.4, before test crashing I can see the GUI popups:
![Screenshot from 2020-04-21 15-49-03](https://user-images.githubusercontent.com/32963518/79867962-ddcb6580-83e7-11ea-98e0-a37fed083cc3.png)
![Screenshot from 2020-04-21 15-49-20](https://user-images.githubusercontent.com/32963518/79867968-e0c65600-83e7-11ea-90d5-46ad506411e8.png)

Should are they in German, no?

MarcoFalke commented:
![Screenshot from 2020-04-21 09-03-06](https://user-images.githubusercontent.com/6399679/79869400-49dda780-83af-11ea-88b1-c0f9fc84c1f7.png)

![Screenshot from 2020-04-21 09-03-09](https://user-images.githubusercontent.com/6399679/79869408-4ba76b00-83af-11ea-8742-cdbad787c6ab.png)


Anyway, if this is not a regression introduced in this pull, it should be reported as a new issue.

hebasto commented:
@MarcoFalke What system are you testing on?

MarcoFalke commented:
Fedora 32

hebasto reviewed (approved):
ACK fa8068acc34f60536f6b55e7326df807f2059a53, tested [manually](https://github.com/bitcoin/bitcoin/pull/18699#pullrequestreview-397221653) and with [scripts](https://github.com/bitcoin/bitcoin/pull/18699#issuecomment-615926628):
- on Fedora 31 (using `LC_ALL=de_DE.UTF-8`):
```
$ XDG_SESSION_TYPE="" QT_QPA_PLATFORM=minimal LC_ALL=de_DE.UTF-8 BITCOIND=/home/hebasto/bitcoin/src/qt/bitcoin-qt ./test/functional/wallet_basic.py
```
- on Ubuntu 18.04.4 (using `LANGUAGE=de`):
```
$ XDG_SESSION_TYPE="" QT_QPA_PLATFORM=minimal LANGUAGE=de BITCOIND=/home/hebasto/bitcoin/src/qt/bitcoin-qt ./test/functional/wallet_basic.py
```

On master (b470c758470fd97ccb42d0348a32942afa2a3ec8) scripts fail with an error:
```
AssertionError: Expected substring not found in error message:
substring: 'Insufficient funds'
error message: 'Ungenügendes Guthaben'.
```

hebasto reviewed (approved):
ACK fa8068acc34f60536f6b55e7326df807f2059a53, tested [manually](https://github.com/bitcoin/bitcoin/pull/18699#pullrequestreview-397221653) and with [scripts](https://github.com/bitcoin/bitcoin/pull/18699#issuecomment-615926628):
- on Fedora 31 (using `LC_ALL=de_DE.UTF-8`):
```
$ XDG_SESSION_TYPE="" QT_QPA_PLATFORM=minimal LC_ALL=de_DE.UTF-8 BITCOIND=/home/hebasto/bitcoin/src/qt/bitcoin-qt ./test/functional/wallet_basic.py
```
- on Ubuntu 18.04.4 (using `LANGUAGE=de`):
```
$ XDG_SESSION_TYPE="" QT_QPA_PLATFORM=minimal LANGUAGE=de BITCOIND=/home/hebasto/bitcoin/src/qt/bitcoin-qt ./test/functional/wallet_basic.py
```

On master (b470c758470fd97ccb42d0348a32942afa2a3ec8) scripts fail with an error:
```
AssertionError: Expected substring not found in error message:
substring: 'Insufficient funds'
error message: 'Ungenügendes Guthaben'.
```

hebasto reviewed (approved):
ACK fa8068acc34f60536f6b55e7326df807f2059a53, tested [manually](https://github.com/bitcoin/bitcoin/pull/18699#pullrequestreview-397221653) and with [scripts](https://github.com/bitcoin/bitcoin/pull/18699#issuecomment-615926628):
- on Fedora 31 (using `LC_ALL=de_DE.UTF-8`):
```
$ XDG_SESSION_TYPE="" QT_QPA_PLATFORM=minimal LC_ALL=de_DE.UTF-8 BITCOIND=/home/hebasto/bitcoin/src/qt/bitcoin-qt ./test/functional/wallet_basic.py
```
- on Ubuntu 18.04.4 (using `LANGUAGE=de`):
```
$ XDG_SESSION_TYPE="" QT_QPA_PLATFORM=minimal LANGUAGE=de BITCOIND=/home/hebasto/bitcoin/src/qt/bitcoin-qt ./test/functional/wallet_basic.py
```

On master (b470c758470fd97ccb42d0348a32942afa2a3ec8) scripts fail with an error:
```
AssertionError: Expected substring not found in error message:
substring: 'Insufficient funds'
error message: 'Ungenügendes Guthaben'.
```

hebasto reviewed (approved):
ACK fa8068acc34f60536f6b55e7326df807f2059a53, tested [manually](https://github.com/bitcoin/bitcoin/pull/18699#pullrequestreview-397221653) and with [scripts](https://github.com/bitcoin/bitcoin/pull/18699#issuecomment-615926628):
- on Fedora 31 (using `LC_ALL=de_DE.UTF-8`):
```
$ XDG_SESSION_TYPE="" QT_QPA_PLATFORM=minimal LC_ALL=de_DE.UTF-8 BITCOIND=/home/hebasto/bitcoin/src/qt/bitcoin-qt ./test/functional/wallet_basic.py
```
- on Ubuntu 18.04.4 (using `LANGUAGE=de`):
```
$ XDG_SESSION_TYPE="" QT_QPA_PLATFORM=minimal LANGUAGE=de BITCOIND=/home/hebasto/bitcoin/src/qt/bitcoin-qt ./test/functional/wallet_basic.py
```

On master (b470c758470fd97ccb42d0348a32942afa2a3ec8) scripts fail with an error:
```
AssertionError: Expected substring not found in error message:
substring: 'Insufficient funds'
error message: 'Ungenügendes Guthaben'.
```

hebasto commented:
@MarcoFalke due to the recent GH glitch my single comment becomes multiple clones; please remove redundant ones.

MarcoFalke commented:
Sorry, had to force push the last commit again to update the tests

hebasto reviewed (approved):
ACK fafca4ad4957defd2aef3f2af07fd3926adbadd8, all unit and functional tests passed locally.

laanwj commented:
Concept ACK, however this adds a fair number of translation messages that shouldn't be there IMO because translators will have difficulty with them due to technicality, ambiguous wording, or spilling internal technical details:

```diff
+ QT_TRANSLATE_NOOP("bitcoin-core", ""
+"Transaction needs a change address, but we can't generate it. Please call "
+"keypoolrefill first."),
+QT_TRANSLATE_NOOP("bitcoin-core", "Error: Keypool ran out, please call keypoolrefill first"),
+QT_TRANSLATE_NOOP("bitcoin-core", "Invalid or non-wallet transaction id"),
+QT_TRANSLATE_NOOP("bitcoin-core", "Transaction has been mined, or is conflicted with a mined transaction"),
+QT_TRANSLATE_NOOP("bitcoin-core", "Transaction has descendants in the mempool"),
+QT_TRANSLATE_NOOP("bitcoin-core", "Transaction has descendants in the wallet"),
```
I think if any of these appear in the GUI they're not helpful in the first place? Do users manually do `keypoolrefill` for example? *mempool* is not a GUI concept at all, it isn't reflected anywhere. What do you have to do if "Invalid or non-wallet transaction id"? (can this even happen through the GUI?)

MarcoFalke commented:
> +"Transaction needs a change address, but we can't generate it. Please call "
+"keypoolrefill first."),


> +QT_TRANSLATE_NOOP("bitcoin-core", "Error: Keypool ran out, please call keypoolrefill first"),


Those are translations already in master.


> +QT_TRANSLATE_NOOP("bitcoin-core", "Invalid or non-wallet transaction id"),

Thanks, added a commit to mark them untranslated.

MarcoFalke commented:
This is the diff I get:

```diff
[src]$ git diff ./qt/bitcoinstrings.cpp
diff --git a/src/qt/bitcoinstrings.cpp b/src/qt/bitcoinstrings.cpp
index ca9546ecb9..9028d3007a 100644
--- a/src/qt/bitcoinstrings.cpp
+++ b/src/qt/bitcoinstrings.cpp
@@ -36,6 +36,10 @@ QT_TRANSLATE_NOOP("bitcoin-core", ""
 "Invalid amount for -maxtxfee=<amount>: '%s' (must be at least the minrelay "
 "fee of %s to prevent stuck transactions)"),
 QT_TRANSLATE_NOOP("bitcoin-core", ""
+"Passphrase provided but private keys are disabled. A passphrase is only used "
+"to encrypt private keys, so cannot be used for wallets with private keys "
+"disabled."),
+QT_TRANSLATE_NOOP("bitcoin-core", ""
 "Please check that your computer's date and time are correct! If your clock "
 "is wrong, %s will not work properly."),
 QT_TRANSLATE_NOOP("bitcoin-core", ""
@@ -118,6 +122,8 @@ QT_TRANSLATE_NOOP("bitcoin-core", "Error: A fatal internal error occurred, see d
 QT_TRANSLATE_NOOP("bitcoin-core", "Error: Disk space is low for %s"),
 QT_TRANSLATE_NOOP("bitcoin-core", "Error: Disk space is too low!"),
 QT_TRANSLATE_NOOP("bitcoin-core", "Error: Keypool ran out, please call keypoolrefill first"),
+QT_TRANSLATE_NOOP("bitcoin-core", "Error: Wallet created but failed to encrypt."),
+QT_TRANSLATE_NOOP("bitcoin-core", "Error: Wallet was encrypted but could not be unlocked"),
 QT_TRANSLATE_NOOP("bitcoin-core", "Failed to listen on any port. Use -listen=0 if you want this."),
 QT_TRANSLATE_NOOP("bitcoin-core", "Failed to rescan the wallet during initialization"),
 QT_TRANSLATE_NOOP("bitcoin-core", "Importing..."),
@@ -182,6 +188,10 @@ QT_TRANSLATE_NOOP("bitcoin-core", "Upgrading txindex database"),
 QT_TRANSLATE_NOOP("bitcoin-core", "User Agent comment (%s) contains unsafe characters."),
 QT_TRANSLATE_NOOP("bitcoin-core", "Verifying blocks..."),
 QT_TRANSLATE_NOOP("bitcoin-core", "Verifying wallet(s)..."),
+QT_TRANSLATE_NOOP("bitcoin-core", "Wallet %s already exists."),
+QT_TRANSLATE_NOOP("bitcoin-core", "Wallet creation failed."),
+QT_TRANSLATE_NOOP("bitcoin-core", "Wallet file verification failed."),
+QT_TRANSLATE_NOOP("bitcoin-core", "Wallet loading failed."),
 QT_TRANSLATE_NOOP("bitcoin-core", "Wallet needed to be rewritten: restart %s to complete"),
 QT_TRANSLATE_NOOP("bitcoin-core", "Warning: unknown new rules activated (versionbit %i)"),
 QT_TRANSLATE_NOOP("bitcoin-core", "Zapping all transactions from wallet..."),

luke-jr commented:
Concept ACK

promag commented:
Concept ACK.

MarcoFalke commented:
Rebased

MarcoFalke commented:
Ok, I marked everything that wasn't translated before as `Untranslated()`. Should be easier to review now.

MarcoFalke commented:
Rebased

hebasto reviewed (approved):
ACK fa2cce4391b0b1bda325f695bb45f7b565c8e8ea

> Ok, I marked everything that wasn't translated before as `Untranslated()`.

Verified by code reviewing that no new translations added.

MarcoFalke commented:
Can also be verified with `(cd src && make translate) && git diff` before and after (diff of diff should be empty). :)

laanwj commented:
ACK fa2cce4391b0b1bda325f695bb45f7b565c8e8ea, checked that no new translation messages are added compared to master.

hebasto commented on src/util/translation.h:
```diff
@@ -18,6 +18,20 @@ struct bilingual_str {
+        lhs.translated + rhs.translated};
+}
+
+/** Mark a bilingual_str as untranslated */
+inline static bilingual_str Untranslated(std::string original) { return {original, original}; }
+/** Unary operator to return the original */
+inline static std::string OpOriginal(const bilingual_str& b) { return b.original; }
+/** Unary operator to return the translation */
+inline static std::string OpTranslated(const bilingual_str& b) { return b.translated; }
```
@MarcoFalke What is the rationale to declare these functions with `static`?

MarcoFalke replied:
I think this is a leftover. They only need to be `inline` to avoid linker errors. `static` might have no effect.

MarcoFalke replied:
Feel free to remove in your pull
